deadlock in tests

```
=== CONT  TestStreamPlacementUpdated/ColdStream
POTENTIAL DEADLOCK:
goroutine 360662 have been trying to lock 0xc015919528 for more than 30s:
  github.com/towns-protocol/towns/core/node/events.(*Stream).AdvanceStickyPeer()
      /home/runner/_work/towns/towns/core/node/events/stream.go:1111 +0x41
  github.com/towns-protocol/towns/core/node/events.(*StreamCache).syncStreamFromPeers()
      /home/runner/_work/towns/towns/core/node/events/stream_sync_task.go:277 +0x871
  github.com/towns-protocol/towns/core/node/events.(*StreamCache).reconciliationTask()
      /home/runner/_work/towns/towns/core/node/events/stream_sync_task.go:154 +0x3d9
  github.com/towns-protocol/towns/core/node/events.(*StreamCache).submitReconciliationTask.func2()
      /home/runner/_work/towns/towns/core/node/events/stream_sync_task.go:118 +0x84
  github.com/gammazero/workerpool.worker()
      /home/runner/go/pkg/mod/github.com/gammazero/workerpool@v1.1.3/workerpool.go:237 +0x35

goroutine 361477 previously locked it from:
  github.com/towns-protocol/towns/core/node/events.(*Stream).lockMuAndLoadView()
      /home/runner/_work/towns/towns/core/node/events/stream.go:127 +0x45
  github.com/towns-protocol/towns/core/node/events.(*Stream).applyStreamMiniblockUpdates()
      /home/runner/_work/towns/towns/core/node/events/stream.go:1023 +0x87
  github.com/towns-protocol/towns/core/node/events.(*StreamCache).onBlockWithLogs.func2()
      /home/runner/_work/towns/towns/core/node/events/stream_cache.go:248 +0x6e4
  github.com/gammazero/workerpool.worker()
      /home/runner/go/pkg/mod/github.com/gammazero/workerpool@v1.1.3/workerpool.go:237 +0x35

goroutine 361477 current stack:
goroutine 361477 [sync.RWMutex.RLock]:
sync.runtime_SemacquireRWMutexR(0xc0130f4440?, 0x1?, 0x2387a35?)
	/home/runner/_work/_tool/go/1.24.2/x64/src/runtime/sema.go:100 +0x25
sync.(*RWMutex).RLock(0xc0130f4430)
	/home/runner/_work/_tool/go/1.24.2/x64/src/sync/rwmutex.go:74 +0x5b
github.com/towns-protocol/towns/core/node/events.(*StreamCache).submitToPool(0xc0130f43c0, 0xc005387d00, 0xc017b486c0)
	/home/runner/_work/towns/towns/core/node/events/stream_sync_task.go:50 +0x45
github.com/towns-protocol/towns/core/node/events.(*StreamCache).submitGetRecordTask(0xc0130f43c0, 0xc005387d00, 0xc015919500)
	/home/runner/_work/towns/towns/core/node/events/stream_sync_task.go:66 +0x1b5
github.com/towns-protocol/towns/core/node/events.(*StreamCache).submitSyncStreamTaskToPool(0xc0130f43c0, 0xc005387d00, 0xc015919500, 0x0)
	/home/runner/_work/towns/towns/core/node/events/stream_sync_task.go:40 +0x65
github.com/towns-protocol/towns/core/node/events.(*StreamCache).SubmitSyncStreamTask(...)
	/home/runner/_work/towns/towns/core/node/events/stream_sync_task.go:31
github.com/towns-protocol/towns/core/node/events.(*Stream).applyStreamMiniblockUpdates(0xc015919500, {0x3453768, 0xc01274e640}, {0xc0022fe8d0, 0x1, 0x3498ee0?}, 0x43)
	/home/runner/_work/towns/towns/core/node/events/stream.go:1053 +0x4d1
github.com/towns-protocol/towns/core/node/events.(*StreamCache).onBlockWithLogs.func2()
	/home/runner/_work/towns/towns/core/node/events/stream_cache.go:248 +0x6e5
github.com/gammazero/workerpool.worker(0xc0180310e0, 0xc0183f1500, 0xc00be5d610)
	/home/runner/go/pkg/mod/github.com/gammazero/workerpool@v1.1.3/workerpool.go:237 +0x36
created by github.com/gammazero/workerpool.(*WorkerPool).dispatch in goroutine 361476
	/home/runner/go/pkg/mod/github.com/gammazero/workerpool@v1.1.3/workerpool.go:197 +0x4bd
Other goroutines holding locks:
goroutine 614759 lock 0xc01d701988
  github.com/towns-protocol/towns/core/node/crypto.(*transactionPool).Submit()
      /home/runner/_work/towns/towns/core/node/crypto/chain_txpool.go:703 +0x66
  github.com/towns-protocol/towns/core/node/registries.(*RiverRegistryContract).SetStreamLastMiniblockBatch()
      /home/runner/_work/towns/towns/core/node/registries/river_registry_contract.go:741 +0x1ec
  github.com/towns-protocol/towns/core/node/events.(*miniblockProducer).submitProposalBatch()
      /home/runner/_work/towns/towns/core/node/events/miniblock_producer.go:414 +0x14c4
  github.com/towns-protocol/towns/core/node/events.(*candidateTracker).add.func1()
      /home/runner/_work/towns/towns/core/node/events/miniblock_producer.go:149 +0x1ac

goroutine 312232 lock 0xc015f9e588
  github.com/towns-protocol/towns/core/node/crypto.(*transactionPool).Submit()
      /home/runner/_work/towns/towns/core/node/crypto/chain_txpool.go:703 +0x66
  github.com/towns-protocol/towns/core/node/rpc.(*serviceTester).startAutoMining.func1()
      /home/runner/_work/towns/towns/core/node/rpc/tester_test.go:273 +0x224

goroutine 582749 lock 0xc0249de540
  github.com/towns-protocol/towns/core/node/crypto.(*chainMonitor).runWithBlockPeriod()
      /home/runner/_work/towns/towns/core/node/crypto/chain_monitor.go:370 +0x172b

goroutine 342350 lock 0xc011b589a0
  github.com/towns-protocol/towns/core/node/crypto.(*chainMonitor).runWithBlockPeriod()
      /home/runner/_work/towns/towns/core/node/crypto/chain_monitor.go:370 +0x172b

goroutine 613901 lock 0xc0041bf2a8
  github.com/towns-protocol/towns/core/node/crypto.(*transactionPool).Submit()
      /home/runner/_work/towns/towns/core/node/crypto/chain_txpool.go:703 +0x66
  github.com/towns-protocol/towns/core/node/registries.(*RiverRegistryContract).AllocateStream()
      /home/runner/_work/towns/towns/core/node/registries/river_registry_contract.go:195 +0x301
  github.com/towns-protocol/towns/core/node/nodes.(*streamRegistryImpl).AllocateStream()
      /home/runner/_work/towns/towns/core/node/nodes/stream_registry.go:66 +0x228
  github.com/towns-protocol/towns/core/node/rpc.(*Service).createReplicatedStream()
      /home/runner/_work/towns/towns/core/node/rpc/create_stream.go:170 +0x2de
  github.com/towns-protocol/towns/core/node/rpc.(*Service).createStream()
      /home/runner/_work/towns/towns/core/node/rpc/create_stream.go:133 +0x112c
  github.com/towns-protocol/towns/core/node/rpc.(*Service).createStreamImpl()
      /home/runner/_work/towns/towns/core/node/rpc/create_stream.go:26 +0x7d
  github.com/towns-protocol/towns/core/node/rpc.executeConnectHandler[...]()
      /home/runner/_work/towns/towns/core/node/rpc/forwarder.go:134 +0x1eb
  github.com/towns-protocol/towns/core/node/rpc.(*Service).CreateStream()
      /home/runner/_work/towns/towns/core/node/rpc/forwarder.go:161 +0x12a
  connectrpc.com/connect.NewUnaryHandler[...].func1()
      /home/runner/go/pkg/mod/connectrpc.com/connect@v1.17.0/handler.go:51 +0x250
  github.com/towns-protocol/towns/core/node/rpc.(*Service).initHandlers.NewTimeoutInterceptor.func1.1()
      /home/runner/_work/towns/towns/core/node/rpc/timeout_interceptor.go:24 +0x11b
  github.com/towns-protocol/towns/core/node/rpc.(*metricsInterceptor).WrapUnary.func1()
      /home/runner/_work/towns/towns/core/node/rpc/metrics_interceptor.go:88 +0x55d
  connectrpc.com/connect.NewUnaryHandler[...].func2()
      /home/runner/go/pkg/mod/connectrpc.com/connect@v1.17.0/handler.go:69 +0x12a
  connectrpc.com/connect.(*Handler).ServeHTTP()
      /home/runner/go/pkg/mod/connectrpc.com/connect@v1.17.0/handler.go:237 +0xd66
  github.com/towns-protocol/towns/core/node/protocol/protocolconnect.NewStreamServiceHandler.func1()
      /home/runner/_work/towns/towns/core/node/protocol/protocolconnect/protocol.connect.go:445 +0x38e
  net/http.HandlerFunc.ServeHTTP()
      /home/runner/_work/_tool/go/1.24.2/x64/src/net/http/server.go:2294 +0x47
  github.com/towns-protocol/towns/core/node/rpc.(*httpHandler).ServeHTTP()
      /home/runner/_work/towns/towns/core/node/rpc/http_handler.go:52 +0xb3e
  net/http.(*ServeMux).ServeHTTP()
      /home/runner/_work/_tool/go/1.24.2/x64/src/net/http/server.go:2[822](https://github.com/towns-protocol/towns/actions/runs/15057716234/job/42326956525?pr=3080#step:16:823) +0x23e
  github.com/towns-protocol/towns/core/node/rpc.(*Service).runHttpServer.(*Cors).Handler.func2()
      /home/runner/go/pkg/mod/github.com/rs/cors@v1.11.1/cors.go:289 +0x2e6
  net/http.HandlerFunc.ServeHTTP()
      /home/runner/_work/_tool/go/1.24.2/x64/src/net/http/server.go:2294 +0x47
  net/http.serverHandler.ServeHTTP()
      /home/runner/_work/_tool/go/1.24.2/x64/src/net/http/server.go:3301 +0x2a1
  net/http.initALPNRequest.ServeHTTP()
      /home/runner/_work/_tool/go/1.24.2/x64/src/net/http/server.go:3974 +0x358
  golang.org/x/net/http2.(*serverConn).runHandler()
      /home/runner/go/pkg/mod/golang.org/x/net@v0.38.0/http2/server.go:2433 +0x253



panic: deadlock detected

goroutine 361424 [running]:
github.com/linkdata/deadlock.(*Options).PotentialDeadlock(...)
	/home/runner/go/pkg/mod/github.com/linkdata/deadlock@v0.5.3/options.go:77
github.com/linkdata/deadlock.(*lockOrder).timeoutFn(0x0?, 0xc013b28d20, 0x6fc23ac00, 0x580d6, {0xc00e89bda0, 0x6, 0x6}, {0x2d766a0, 0xc015919528})
	/home/runner/go/pkg/mod/github.com/linkdata/deadlock@v0.5.3/lockorder.go:144 +0x4c5
created by github.com/linkdata/deadlock.lock in goroutine 360662
	/home/runner/go/pkg/mod/github.com/linkdata/deadlock@v0.5.3/lock.go:25 +0x3e7
FAIL	github.com/towns-protocol/towns/core/node/rpc	93.251s
```

--- a/core/node/events/stream.go
+++ b/core/node/events/stream.go
@@ -1130,6 +1130,7 @@ func (s *Stream) applyStreamMiniblockUpdates(
 
 	view, err := s.lockMuAndLoadView(ctx)
 	if err != nil {
+		s.mu.Unlock()
 		logging.FromCtx(ctx).Errorw("applyStreamEvents: failed to load view", "error", err)
 		return
 	}
@@ -1174,9 +1175,8 @@ func (s *Stream) applyStreamMiniblockUpdates(
 	s.lastAppliedBlockNum = blockNum
 	s.mu.Unlock()
 
-	// Submit sync task after releasing the lock to avoid deadlock
 	if needsSyncTask {
-		s.params.streamCache.SubmitSyncStreamTask(s, nil)
+		s.params.streamCache.SubmitReconcileStreamTask(s, nil)
 	}
 }
 

--- a/core/node/events/stream.go
+++ b/core/node/events/stream.go
@@ -1129,15 +1129,18 @@ func (s *Stream) applyStreamMiniblockUpdates(
 	}
 
 	view, err := s.lockMuAndLoadView(ctx)
-	defer s.mu.Unlock()
 	if err != nil {
 		logging.FromCtx(ctx).Errorw("applyStreamEvents: failed to load view", "error", err)
 		return
 	}
 
 	if view == nil {
+		s.mu.Unlock()
 		return // stream is not local, no need to apply miniblock updates
 	}
+
+	// Track if we need to submit a sync task after releasing the lock
+	needsSyncTask := false
 
 	// TODO: REPLICATION: FIX: this function now can be called multiple times per block.
 	// Sanity check
@@ -1158,7 +1161,7 @@ func (s *Stream) applyStreamMiniblockUpdates(
 			})
 			if err != nil {
 				if IsRiverErrorCode(err, Err_STREAM_RECONCILIATION_REQUIRED) {
-					s.params.streamCache.SubmitReconcileStreamTask(s, nil)
+					needsSyncTask = true
 				} else {
 					logging.FromCtx(ctx).Errorw("onStreamLastMiniblockUpdated: failed to promote candidate", "error", err)
 				}
@@ -1169,6 +1172,12 @@ func (s *Stream) applyStreamMiniblockUpdates(
 	}
 
 	s.lastAppliedBlockNum = blockNum
+	s.mu.Unlock()
+
+	// Submit sync task after releasing the lock to avoid deadlock
+	if needsSyncTask {
+		s.params.streamCache.SubmitSyncStreamTask(s, nil)
+	}
 }
 
 // GetQuorumNodes returns the list of nodes this stream resides on according to the stream

